### DISCUSSION TITLE
Access to QuestPdfSkia.dll is denied

### DISCUSSION THREAD

**Discussion by jplachy1** [QuestPDF/QuestPDF#1009] [Q&A] [answered]
Im using QuestPdf to generate pdfs in my MVC app running on .NET Framework  4.8. Everything worked fine until i've deployed the app on the staging server.

First deployment was fine, but after QuestPdfSkia.dll appears in the site_root\bin\QuestPdfSkia.dll other deployments fail and i get the rror "Access to QuestPdfSkia.dll is denied". I have found out that this file is locked by the IIS Worker Process, so if i want to deploy I have to stop the app in the IIS and then deploy.

I can solve this by creating a script that first stops the app and then deploys it, but it's odd that this problem occured never before with any other .dll file (the app is being developed for 10 years.) and im not sure if this is the right approach.

I have also tried to place the `app_offline.htm` file into the site root directory which should apparently release the locks a not respond to any incoming http requests which happens, but I still am unable to deploy and i get the same error as before "Access to QuestPdfSkia.dll is denied".

Has anyone had a similiar problem?

**Comment by sharbhajan** [comment]
I'm facing the same problem. I have more than one .NET Framework 4.8 ASP.NET application running under the same application pool. Every application has its own bin folder with QuestPdfSkia.dll in each of them. But still while updating an application this file is locked and unable to replace. Could it be due to shared application pool?
The problem started only a few weeks ago, it was working fine earlier. I've to update only some particular applications frequently while others are rarely updated. And these updates are mostly unrelated to pdf generation. Occasionally, restarting only application pool works, but sometimes IIS needs to be restarted. Restarting application pool or IIS takes down other applications.

OS: Windows Server 2019
IIS: 10.0
QuestPDF: 2024.7.3

**Reply by jplachy1** [reply]
I solved this by creating custom deploy script that stops the application pool then I deploy the app and then start the application pool again.  We've been using this solution in production for 6 months and everything works fine. I also had to include the QuestPdfSkia.dll file in my application source code, I don't remember why exaclty but i know that I had to do it.

Here is the snippet of stopping and starting the app pool.

```
function Suspend-Site {

    [CmdletBinding(SupportsShouldProcess)]
    param(
        $target,
        $appOffline)

    $deployURL = Get-Deploy-Url($target)
    $siteName = $target.SiteName

    Write-Log-Site-Action $target "Taking site offline"

    if($PSCmdlet.ShouldProcess("$target")) {

        & $MsDeploy -verb:sync -source:recycleApp -dest:recycleApp="$siteName",recycleMode="StopAppPool",computerName="$deployURL",userName="$user",password="$password",authType="Basic",includeAcls="False" -allowUntrusted
        if($LASTEXITCODE -ne 0) {
            Write-Log-Site-Error $target "Stopping application pool failed"
            exit 1;
        }

        & $MsDeploy -verb:sync -source:contentPath="$appOffline" -dest:contentPath="$siteName/app_offline.htm",computerName="$deployURL",userName="$user",password="$password",authType="Basic",includeAcls="False" -allowUntrusted

        if($LASTEXITCODE -ne 0) {
            Write-Log-Site-Error $target "Taking site offline failed"
            exit 1;
        }
    }
}

function Resume-Site {

    [CmdletBinding(SupportsShouldProcess)]
    param($target)

    $deployURL = Get-Deploy-Url($target)
    $siteName = $target.SiteName

    Write-Log-Site-Action $target "Taking site online"

    if($PSCmdlet.ShouldProcess("$target")) {
        & $MsDeploy -verb:sync -source:recycleApp -dest:recycleApp="$siteName",recycleMode="StartAppPool",computerName="$deployURL",userName="$user",password="$password",authType="Basic",includeAcls="False" -allowUntrusted
        if($LASTEXITCODE -ne 0) {
            Write-Log-Site-Error $target "Starting application pool failed"
            exit 1;
        }

        & $MsDeploy -verb:delete -dest:contentPath="$siteName/app_offline.htm",computerName="$deployURL",userName="$user",password="$password",authType="Basic",includeAcls="False" -allowUntrusted

        if($LASTEXITCODE -ne 0) {
            Write-Log-Site-Error $target "Taking site online failed"
            exit 1;
        }
    }
}

```

**Comment by MarcinZiabek** [comment]
I am very sorry for this experience. The described problem indeed seems to be problematic 😞 

As far as I understand, this is a known problem with IIS / .NET Framework. I remember experiencing similar issues with SQL Server DLLs around 8 years ago, but I haven't found a solution.

Moreover, I am unaware of any recent changes that could cause this issue.

**Reply by sharbhajan** [reply]
Hi @MarcinZiabek some of my apps are 6 years old, others are upto 15 years old (updated and developing) and till date none of the other packages or my app files got locked while updating. QuestPdfSkia.dll is the only one getting locked in my case.

**Comment by MarcinZiabek** [comment, marked as answer]
Closing this discussion with a summary, since the root cause is well understood and both reporters found working solutions.

What's happening: QuestPdfSkia.dll is a native library. Managed assemblies can be replaced during deployment because ASP.NET loads them through a shadow-copy mechanism that leaves the original files unlocked. Native libraries work differently: they are loaded directly from disk by the operating system and remain locked by the worker process for as long as it runs. This is also why app_offline.htm doesn't help — it recycles the application, but native libraries are only released when the process itself exits.

This is not specific to QuestPDF — the same applies to any native dependency under IIS (SkiaSharp, SQLite, and others).

Recommended solution: stop the application pool before deployment and start it afterwards, exactly as @jplachy1 shared above (thank you for the script!). This is standard practice for deploying apps with native dependencies on IIS.
